core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/pythonUDF/PythonUDFOpExec.java

```diff
@@ -441,69 +351,230 @@ private static void readArrowStream(FlightClient client, String descriptorPath,
                 convertArrow2AmberTableBuffer(root, resultQueue);
                 root.clear();
             }
-        } catch (Exception e) {
+        } catch (RuntimeException e) {
             closeAndThrow(client, e);
         }
     }
 
     /**
-     * Make the execution of the UDF in Python and read the results that are passed back. This should only be called
-     * after input data is passed to Python. This is a blocking call.
+     * Converts an Amber schema into Arrow schema.
      *
-     * @param client      The FlightClient that manages this.
-     * @param mapper      Used to decode the result status message (Json).
-     * @param resultQueue To store the results. Must be empty when it is passed here.
+     * @param amberSchema The Amber Tuple Schema.
+     * @return An Arrow {@link org.apache.arrow.vector.types.pojo.Schema}.
      */
-    private static void executeUDF(FlightClient client, ObjectMapper mapper, Queue<Tuple> resultQueue) {
+    private static org.apache.arrow.vector.types.pojo.Schema convertAmber2ArrowSchema(Schema amberSchema)
```

Comment:
I assume all these changes are just reordering of the original code

Reply:
most changes here are due to reordering. but some noticeable changes:
1. `InputExhausted` is now forwarded to Python to inform the end of input. This is useful for training operators who consume every tuple and not doing the actual training/testing logic until input is exhausted.
2. The Python process is now forced to flush log with `-u` argument, so that the python’s IO can be inherited by the Java side.
3. The Python process is killed after notifying peaceful termination, just to double make sure the process is terminated.